internal/receiver/discoveryreceiver/correlation.go

```diff
@@ -0,0 +1,217 @@
+type correlationStore interface {
+	UpdateEndpoint(endpoint observer.Endpoint, state endpointState, observerID config.ComponentID)
+	GetOrCreate(receiverID config.ComponentID, endpointID observer.EndpointID) correlation
+	Attrs(receiverID config.ComponentID) map[string]string
+	UpdateAttrs(receiverID config.ComponentID, attrs map[string]string)
+	// Start the reaping loop to prevent unnecessary endpoint buildup
+	Start()
+	// Stop the reaping loop
+	Stop()
+}
+
+// store is a collection of mappings used as an instantaneous record of
+// 1. endpoints to their associated receivers->correlations
+// 2. receivers to their endpoint-agnostic Attrs used as a message
+// passing mechanism (currently just for embedded config values).
+// This way the pre-created receiver instances can have attrs
+// before they are created via endpoint.
+type store struct {
+	logger *zap.Logger
+	// ~map[endpointID]map[receiverID]*corr
+	correlations  *sync.Map
+	endpointLocks *keyLock
+	receiverAttrs *sync.Map
+	receiverLocks *keyLock
+	sentinel      chan struct{}
+	reapInterval  time.Duration
+	ttl           time.Duration
+}
+
+func newCorrelationStore(logger *zap.Logger, ttl time.Duration) correlationStore {
+	return &store{
+		logger:        logger,
+		correlations:  &sync.Map{},
+		endpointLocks: newKeyLock(),
+		receiverAttrs: &sync.Map{},
+		receiverLocks: newKeyLock(),
+		reapInterval:  30 * time.Second,
+		ttl:           ttl,
+		sentinel:      make(chan struct{}, 1),
+	}
+}
+
+// UpdateEndpoint will update all existing correlation timestamps and states by endpoint.ID, or
+// creates a new no-type ~singleton w/ the initial correlation info for later use in correlation creation.
+func (s *store) UpdateEndpoint(endpoint observer.Endpoint, state endpointState, observerID config.ComponentID) {
+	defer s.endpointLocks.Lock(endpoint.ID)()
+	receiverStore, ok := s.correlations.LoadOrStore(endpoint.ID,
+		func() *sync.Map {
+			receiverMap := &sync.Map{}
+			receiverMap.Store(statussources.NoType, &correlation{
+				endpoint:    endpoint,
+				observerID:  observerID,
+				lastState:   state,
+				lastUpdated: time.Now(),
+			})
+			return receiverMap
+		}())
+	if !ok {
+		// we've set NoType correlation, which is all we can do
+		return
+	}
+	receiverStore.(*sync.Map).Range(func(_, c any) bool {
+		corr := c.(*correlation)
+		corr.endpoint = endpoint
+		// set here for unlikely out of order GetOrCreate eventual consistency
+		corr.observerID = observerID
+		corr.lastUpdated = time.Now()
+		corr.lastState = state
+		return true
+	})
+}
+
+// GetOrCreate returns an existing receiver/endpoint correlation or creates a new one
+// based on the no-type ~singleton for the last endpoint update event.
+func (s *store) GetOrCreate(receiverID config.ComponentID, endpointID observer.EndpointID) correlation {
+	endpointUnlock := s.endpointLocks.Lock(endpointID)
+	rStore, _ := s.correlations.LoadOrStore(endpointID,
+		func() *sync.Map {
```

Comment:
Can you help me understand the need for anonymous function here?

Reply:
Wondering if it would help readers to describe the concurrency strategy here.

Reply:
There isn't a need and it was organizational overkill on my part during a refactor pass when changing the store signature. Simplified it a bit.